go-controller/pkg/node/node_smartnic.go

```diff
@@ -0,0 +1,203 @@
+			klog.Infof("DeleteFunc for POD: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
+			if _, ok := servedPods.Load(pod.UID); !ok {
+				return
+			}
+			servedPods.Delete(pod.UID)
+			retryPods.Delete(pod.UID)
+			vfRepName, err := n.getVfRepName(pod)
+			if err != nil {
+				klog.Errorf("Failed to get VF Representor Name from Pod: %s. Representor port may have been deleted.", err)
+				return
+			}
+			err = n.delRepPort(vfRepName)
+			if err != nil {
+				klog.Infof("Failed to delete VF representor %s. %s", vfRepName, err)
+			}
+		},
+	}, nil)
+}
+
+// getVfRepName returns the VF's representor of the VF assigned to the pod
+func (n *OvnNode) getVfRepName(pod *kapi.Pod) (string, error) {
+	smartNicCD := util.SmartNICConnectionDetails{}
+	if err := smartNicCD.FromPodAnnotation(pod.Annotations); err != nil {
+		return "", fmt.Errorf("failed to get smart-nic annotation. %v", err)
+	}
+	return util.GetSriovnetOps().GetVfRepresentorSmartNIC(smartNicCD.PfId, smartNicCD.VfId)
+}
+
+// addRepPort adds the representor of the VF to the ovs bridge
+func (n *OvnNode) addRepPort(pod *kapi.Pod, vfRepName string, ifInfo *cni.PodInterfaceInfo) error {
+	klog.Infof("addRepPort: %s", vfRepName)
+	smartNicCD := util.SmartNICConnectionDetails{}
+	if err := smartNicCD.FromPodAnnotation(pod.Annotations); err != nil {
+		return fmt.Errorf("failed to get smart-nic annotation. %v", err)
+	}
+
+	err := cni.ConfigureOVS(context.TODO(), pod.Namespace, pod.Name, vfRepName, ifInfo, smartNicCD.SandboxId)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Port %s added to bridge br-int", vfRepName)
+
+	link, err := util.GetNetLinkOps().LinkByName(vfRepName)
+	if err != nil {
+		// Note(adrianc): we are lenient with cleanup in this method as pod is going to be retried anyway.
+		_ = n.delRepPort(vfRepName)
+		return fmt.Errorf("failed to get link device for interface %s", vfRepName)
+	}
+
+	klog.Infof("addRepPort: set link mtu %s", vfRepName)
+	if err = util.GetNetLinkOps().LinkSetMTU(link, ifInfo.MTU); err != nil {
+		_ = n.delRepPort(vfRepName)
+		return fmt.Errorf("failed to setup representor port. failed to set MTU for interface %s", vfRepName)
+	}
+
+	klog.Infof("addRepPort: set link up for %s", vfRepName)
+	if err = util.GetNetLinkOps().LinkSetUp(link); err != nil {
+		_ = n.delRepPort(vfRepName)
+		return fmt.Errorf("failed to setup representor port. failed to set link up for interface %s", vfRepName)
+	}
+
+	// Update connection-status annotation
+	// TODO(adrianc): we should update Status in case of error as well
+	connStatus := util.SmartNICConnectionStatus{Status: util.SmartNicConnectionStatusReady, Reason: ""}
+	podAnnotator := kube.NewPodAnnotator(n.Kube, pod)
+	err = connStatus.SetPodAnnotation(podAnnotator)
+	if err != nil {
+		// we should not get here
+		_ = util.GetNetLinkOps().LinkSetDown(link)
+		_ = n.delRepPort(vfRepName)
+		return fmt.Errorf("failed to setup representor port. failed to set pod annotations. %v", err)
+	}
+
```

Comment:
maybe put Line 163-167 in defer it duplicate in171-176

Reply:
I originally considered that, however i didnt see much use of defer bering leveraged for (non trivial) cleanup  in non test packages of the project.

doing it proper we should defer multiple methods  i.e:
at L137 - `defer func() {if err != nil {n.delRepPort()}}`
at L157 - `defer func() {if err != nil {util.GetNetLinkOps.LinkSetDown(link)}}`

Reply:
sticking to original impl here for now.